apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,1350 @@
+        require(_how.length > 0, ERROR_ACL_SIGNER_MISSING);
+        require(_how[0] < 2**160, ERROR_ACL_SIGNER_NOT_ADDRESS);
+
+        address signer = address(_how[0]);
+        (, bool mustSign) = _getSigner(signer);
+        return !mustSign;
+    }
+
+    /**
+    * @dev Tell whether an address can challenge an action
+    * @param _actionId Identification number of the action to be queried
+    * @param _challenger Address of the challenger
+    * @return True if the challenger can be challenge the action, false otherwise
+    */
+    function canPerformChallenge(uint256 _actionId, address _challenger) external view returns (bool) {
+        Action storage action = _getAction(_actionId);
+        return _canPerformChallenge(action.disputable, _challenger);
+    }
+
+    /**
+    * @dev Tell whether an action can be challenged
+    * @param _actionId Identification number of the action to be queried
+    * @return True if the action can be challenged, false otherwise
+    */
+    function canChallenge(uint256 _actionId) external view returns (bool) {
+        Action storage action = _getAction(_actionId);
+        return _canChallenge(_actionId, action);
+    }
+
+    /**
+    * @dev Tell whether an action can be closed.
+    * @dev An action can be closed if it is allowed to:
+    * @dev  - Proceed in the context of this Agreement (see `_canProceed()`)
+    * @dev  - Be closed in the context of the originating disputable app
+    * @param _actionId Identification number of the action to be queried
+    * @return True if the action can be closed, false otherwise
+    */
+    function canClose(uint256 _actionId) external view returns (bool) {
+        Action storage action = _getAction(_actionId);
+        return _canClose(_actionId, action);
+    }
+
+    /**
+    * @dev Tell whether an action can be settled
+    * @param _actionId Identification number of the action to be queried
+    * @return True if the action can be settled, false otherwise
+    */
+    function canSettle(uint256 _actionId) external view returns (bool) {
+        (, Challenge storage challenge, ) = _getChallengedAction(_actionId);
+        return _canSettle(_actionId, challenge);
+    }
+
+    /**
+    * @dev Tell whether an action can be disputed
+    * @param _actionId Identification number of the action to be queried
+    * @return True if the action can be disputed, false otherwise
+    */
+    function canDispute(uint256 _actionId) external view returns (bool) {
+        (, Challenge storage challenge, ) = _getChallengedAction(_actionId);
+        return _canDispute(_actionId, challenge);
+    }
+
+    /**
+    * @dev Tell whether an action's current challenge settlement can be claimed
+    * @param _actionId Identification number of the action to be queried
+    * @return True if the action settlement can be claimed, false otherwise
+    */
+    function canClaimSettlement(uint256 _actionId) external view returns (bool) {
```

Comment:
⚠️ Not sure, but wouldn't this be a problem for multiple challenges? It would be impossible to recover funds from earlier challenges once new challenges got created?

Reply:
One of the main assumptions of the Agreement app is that there can only be one challenge at a time, therefore you will need to close a previous challenge to challenge an action again, i.e. claiming the settlement in this case

Reply:
Actually, the settlement claim can only happen once, right? Because if a challenge is settled the action is closed, and therefore cannot be challenged again.

Reply:
Yes, you are right, if there are multiple challenges, it means none of the previous ones were settled or ruled in favor of the challenger (i.e. the submitter won)

Reply:
Yes, both are true!